### ISSUE TITLE
Add Explore Hierarchy widget to Concept and Place pages (from 1723)

### ISSUE THREAD

**Issue by gigamorph** [project-lux/lux-frontend#25]
**Problem Description:**
The current implementation of the entity hierarchy is difficult to understand and does not make complex relationships clear. 

July 2023: [Teams Chat](https://teams.microsoft.com/l/message/19:GmLeQu1cBR40mfrno-l8lngg7RFCcJs8fvFIw-MJjIM1@thread.tacv2/1690813321803?tenantId=dd8cbebb-2139-4df8-b411-4e3e87abeb5c&groupId=4620eac6-0955-4c69-93e8-b493b21f0df6&parentMessageId=1690813321803&teamName=MarkLogic&channelName=General&createdTime=1690813321803&allowXTenantAccess=false)

**Expected Behavior/Solution:**
There will be a React Flow visualization of the hierarchy based on the below wireframes.
Note relationship with #84  (list view updates)

**Requirements:**

- The component will not render if there is no data
- Initial view: show current entity in the center with the parents rendering to the left of it and children to the right
    - Each parent and child list will render 5 results
    - Each parent and child list will have their own Show More (+5) and Show Less (-5) button if there are more or less results
    - The component will utilize HAL links to retrieve the children (currently implemented as `placeParts` and `conceptChidlren`, respectively)
    - Parents will be rendered from data in the record (`part_of` property in Places and `broader` property in Concepts)
    - The Show More button will not render if there are not more results
    - The Show Less button will not render if there are not less results
- Each entity in the hierarchy will be text, not a link
    - Each entity will have a vertical 3 dot menu button that contains the following:
        - View this Record button that will take the user to the corresponding entity's record page
        - Explore this Record Hierarchy button that will make the corresponding entity's hierarchy render in the component (parents and children of the selected entity will render and the selected entity will be moved to the center of the hierarchy)
- If a user is navigating the hierarchy and is not on the current page's hierarchy there will be a button that will take them back to their initial view/the current page's hierarchy
- There will be an button for viewing the hierarchy in full screen
- There will be a button for toggling back and forth between the list view and the chart view/visualization
- The default view will be the visualization

Needed:

- [x] Wireframe/Mockup - Mike
- [ ] ~Committee discussions - Sarah~
- [ ] ~Feasibility/Team discussion - Sarah~
- [x] Backend requirements - Kam, middle tier ticket linked below
- [x] Frontend requirements - Kam, Mike
- [x] Questions
    - [x] how best to explore hierarchies for Place and Concepts?

**UAT/LUX Examples:**
- https://lux.collections.yale.edu/view/concept/e22c096f-d5a9-470f-a8c5-ca49277ff465 has 59 grandparents And 137 unique great grand parents
- https://lux.collections.yale.edu/view/place/3eee943e-cd9f-4a3b-808c-95d50e1f0eb3
- https://lux.collections.yale.edu/view/place/c86331f7-f1eb-4b4b-8384-ebfa5a29be73
- https://lux.collections.yale.edu/view/concept/b60f4df3-ecef-4d26-9ea5-157170a9ad25 
- https://lux.collections.yale.edu/view/concept/4aa074f3-00ae-438d-aa8d-917fee3acae6 
- https://lux.collections.yale.edu/view/concept/0559cde4-9022-4a4b-8ebb-bd218d3cf734 
- https://lux-front-dev.collections.yale.edu/view/place/7c3652e3-a5d8-45c9-ac11-19e1430554e4

**Dependencies/Blocks:**
- Blocking: https://git.yale.edu/lux-its/lux-backend/issues/72 
    - The current hierarchies are dependent on the code
    - This widget will remove the dependency on the middle tier
- should be done at the same time as: https://github.com/project-lux/lux-frontend/issues/84

**Related Github Issues:**
- Original Github issue: https://git.yale.edu/lux-its/lux-web/issues/1723

**Related Links:**
- BugHerds:
    - https://www.bugherd.com/projects/284041/tasks/2098
    - https://www.bugherd.com/projects/284041/tasks/2077
    - https://www.bugherd.com/projects/284041/tasks/1696
    - https://www.bugherd.com/projects/284041/tasks/1533
    - https://www.bugherd.com/projects/284041/tasks/1532
    - https://www.bugherd.com/projects/284041/tasks/1531

**Wireframes:**
4/16 version:
Note: on hover, the icon will probably be outlined or highlighted in some way as a visual indicator that it is a button. Once the user clicks on it, that corresponding entity will become the center of the hierarchy as previously discussed.

![image](https://github.com/project-lux/lux-frontend/assets/140099762/049d75e1-6ca3-4e07-8602-6b07ed1edf8a)

**Comment by roamye** [comment]
@kamerynB @miketullo95 
Hey Kam. I remember a couple of meetings ago you had a chart where a user could click on the child and parent. Mike was wondering if he could have a screenshot of that to reference for these mockups.

**Comment by kamerynB** [comment]
Old wireframes:
Initial view:
![image](https://github.com/project-lux/lux-frontend/assets/160149818/772f9599-ce21-4f1c-87bb-aaea0904e67e)

Showing expanded grandparents:
![image](https://github.com/project-lux/lux-frontend/assets/160149818/184ad5b4-292a-4d8a-a996-ca282cbb99ff)

Showing expanded siblings 
![image](https://github.com/project-lux/lux-frontend/assets/160149818/421a16ff-4feb-4374-8695-2e4ea38d4b07)

**Comment by miketullo95** [comment]
@prowns @kamerynB @roamye wireframes uploaded. due to this tickets complexity, this may need to be discussed as a group.

**Comment by prowns** [comment]
@miketullo95 and @kamerynB - Some feedback from R/S/J review of the mocks...

- The 3 vertical dots look like a drag handle
- Suggestion to remove the intermediary "View Record/Explore this Hierarchy' selection. 
-- Clicking the name takes user to new page (follows LUX name-clicking convention)
-- Add an icon (TBD) to let people explore in situ

We would need a new icon for this.

**Comment by kamerynB** [comment]
Why not just do a dropdown button similar to the advanced search button on the landing page. The three dots would be an arrow.

![Screenshot 2024-04-10 at 3 06 53 PM](https://github.com/project-lux/lux-frontend/assets/140102078/f4ee31af-dd35-4984-99df-9eb26a256690)

**Comment by prowns** [comment]
I don't mind that, but the goal was eliminating that extra click.  That said, I like the arrow better than three dots.  On the left side, should the arrow be on the left, with the box expanding out. This also implies "up" the hierarchy.

Labels would need to change.  How about: View Record, Explore Hierarchy

**Comment by roamye** [comment]
There are two options up for discussion: 
Option 1: The name of the entity is a link to that entity page and the arrow is a button to a menu that allows a user to view hierarchy OR it's a button that takes the user to that entity's hierarchy. (whichever we choose) 
![image](https://github.com/project-lux/lux-frontend/assets/163148482/bf6e001a-f5f4-45cd-b806-c4638a977bd9)
![image](https://github.com/project-lux/lux-frontend/assets/163148482/3ff755ee-2137-4335-8bd3-a82c7ba0d8be)

Option 2: The text is a button that shows a dropdown menu for user to view options for that entity
![image](https://github.com/project-lux/lux-frontend/assets/163148482/d2e1d7e9-d3e6-4bf7-a0c0-db7c0ae414c5)
![image](https://github.com/project-lux/lux-frontend/assets/163148482/8fea394a-6347-47ba-aab5-d2e26a24eb7b)

After it has been discussed I will move the mockups to the issue body. 
[Conversation in Teams](https://teams.microsoft.com/l/message/19:GmLeQu1cBR40mfrno-l8lngg7RFCcJs8fvFIw-MJjIM1@thread.tacv2/1712946522008?tenantId=dd8cbebb-2139-4df8-b411-4e3e87abeb5c&groupId=4620eac6-0955-4c69-93e8-b493b21f0df6&parentMessageId=1712946522008&teamName=MarkLogic&channelName=General&createdTime=1712946522008 )

**Comment by azaroth42** [comment]
Option 1:
![25-diagram](https://github.com/project-lux/lux-frontend/assets/871868/74804926-0900-45e3-8b68-a16c56a43022)

Option 2:
![25-version2](https://github.com/project-lux/lux-frontend/assets/871868/f1602e9f-55e9-4740-8477-97a024ccd8bb)

**Comment by prowns** [comment]
I like option 2.  :)

**Comment by prowns** [comment]
Older mocks here for reference:
Updated wireframes to be uploaded by @miketullo95.
Default view 
![25 - closed](https://github.com/project-lux/lux-frontend/assets/98550395/a7e4670f-3762-4661-9e95-881add997196)

When a user clicks the menu icon, shows two options.
![25 - open](https://github.com/project-lux/lux-frontend/assets/98550395/d78a430e-0fbc-4c68-8fdd-2e49435b5b81)

When user clicks "show more" it shows 5 more items (placeholder data, not for UAT)
![25 - show less](https://github.com/project-lux/lux-frontend/assets/98550395/c025ce50-4948-4f27-b761-5d6efba82e50)

When a user clicks on "Explore Hierarchy" (and when an item has no parents)
![25 - new record](https://github.com/project-lux/lux-frontend/assets/98550395/064cf785-a444-4241-9a32-3eecf464ffa2)

**Comment by kamerynB** [comment]
Update 5/2/24:
The hierarchy is not in DEV yet but the following is now working locally:
1. Show More/Less in List view and Chart view
2. Connectivity between the List and Chart, ie when selecting to view an entity's hierarchy or clicking show more/less will be reflected in both views when switching back and forth
3. Show all X children results button
4. Show entity's hierarchy button
5. Entity links
6. Fullscreen view with some minor bugs

cc: @prowns @jffcamp @roamye

**Comment by kamerynB** [comment]
Update 5/8:
The first iteration of the hierarchy has been deployed to DEV. Please feel free to leave feedback here.
@prowns @jffcamp @roamye

**Comment by kamerynB** [comment]
@jffcamp Is the below screenshot what you see? And if so, is it the `visual works by material or technique` record that is weird? What I'm thinking is if I set a width on each node, then that might mitigate long titles from rendering weird edges (edges being the lines connecting the nodes/entity links).

![Screenshot 2024-05-16 at 1 56 27 PM](https://github.com/project-lux/lux-frontend/assets/140102078/2f327074-342a-4f45-b189-727247d3af4d)

**Comment by jffcamp** [comment]
Yes. It is the Visual Works by material or technique.

**Comment by prowns** [comment]
Moving visualization around drops artifacts:
![image](https://github.com/project-lux/lux-frontend/assets/140099762/fe82538e-ca8d-44bc-a067-493952ada2c0)

**Comment by roamye** [comment]
The record Algeria in DEV: https://lux-front-dev.collections.yale.edu/view/place/b544756d-3e3c-4a88-8a0c-f5a26b22b353, does not accurately reflect the children when selecting the branch of the child "Constantine (Province)". This happens within the list view and the hierarchy view. Examples below: 

List view of Constantine (Province): 
![image](https://github.com/project-lux/lux-frontend/assets/163148482/78bb65f9-6b15-4cab-8874-5d387174a27e)

Hierarchy View of Constantine (Province): 
![image](https://github.com/project-lux/lux-frontend/assets/163148482/a2f18600-879c-4fb9-9139-aa29a8ca47e4)

What should show for the record Constantine (Province): 
![image](https://github.com/project-lux/lux-frontend/assets/163148482/e6f1bf11-df50-4c9e-b8ee-9eb8b4a85deb)

Note: I clicked on Constantine (Province) record from the Algeria record and it showed as the two photos above rather than immediately as the third photo. I had to refresh it for it to reflect it properly.

**Comment by kamerynB** [comment]
@roamye Thank you for pointing out. The next deployment to DEV, which is TBD, will contain the fix.

**Comment by kamerynB** [comment]
@roamye I have to work on accessibility and overall functionality. It's not a facets issue since the children results come from a search link HAL link. I'll add it to the 8/19 milestone. It won't be done for the current milestone.

**Comment by kamerynB** [comment]
@jffcamp @prowns What is the urgency of this ticket? I have not been working on this as other issues have taken priority. Is this something I should return to for the next deployment?

cc: @roamye

**Comment by jffcamp** [comment]
@kamerynB, it would be great to get this critical ticket completed. So yes, if you can work on it for the next deployment that would be swell.

**Comment by kamerynB** [comment]
Making notes in response to Amy's comment.

Feedback on Functionality:
1. Yes, known issue. Thank you for recording it!

General Feedback:
1. This is because `Uruguay` is a child and parent of the entity but it does not appear in the first 5 children being rendered.
2. It technically does. `Brazil` is a parent and child but it is not listed in the first 5 parents rendered. When you hit Show More, it becomes visible to the frontend and therefore the (poorly placed) connection is made.
3. I have yet to figure out how to make ReactFlow render the nodes in a consistent manner. Right now, it is choosing the coordinates of the nodes based on `dagre`, another Javascript library.

Notes:
Issues 1 and 2 of the General Feedback are because a parent is a child or a child is a parent and they are not being rendered at first because they are not part of the initial 5 parents/children displayed. This should be discussed at the 10/18/24 team meeting.

cc: @roamye

**Comment by roamye** [comment]
From IT Team Meeting 10/18 ([notes here](https://docs.google.com/document/d/1vch7ljI3Wqan-HDfTQs8Ii3vHQRISnrGS-TSyTilWEo/edit?tab=t.0)):
- This issue will move to the next milestone (11/04)  to allow for fixes and testing to be done.
- Kam to include fixes to show instances of hierarchy on both sides when they are categorized as parents & children.

**Comment by azaroth42** [comment]
Areas where I think it's good:

* Loading a new page from zero (hard refresh) the layout is good
* General functionality is good

Areas where I think it still needs a bit of work are mostly the interactions where it redraws after the initial state:

* When you click the trident icon to move between concepts within the viz, the result is much worse than going to the page via a hard refresh, both for the layout and the zoom
* When you click the box icon, the show all x children can disappear off the bottom of the widget
* When you click the trident, the zoom sticks, meaning that the redraw can put all of the content off the visible part of the widget

**Comment by roamye** [comment]
[teams](https://teams.microsoft.com/l/message/19:351e0281aaa144d0812b960d216958c6@thread.tacv2/1731515809216?tenantId=dd8cbebb-2139-4df8-b411-4e3e87abeb5c&groupId=4620eac6-0955-4c69-93e8-b493b21f0df6&parentMessageId=1731515809216&teamName=ML%2F%20Cultural%20Heritage%20Support&channelName=1%20-%20LUX_ML&createdTime=1731515809216) update 11/23: 
Vite migration ( https://github.com/project-lux/lux-frontend/issues/412 ) must happen first before this issue can be deployed.

**Comment by roamye** [comment]
This is being pushed to the 12-02 milestone as: 
- They require certain dependencies to be updated, which is fine. There needs to be updates. However...
- This should come into play when migrating to Vite.

[teams convo](https://teams.microsoft.com/l/message/19:351e0281aaa144d0812b960d216958c6@thread.tacv2/1731692568019?tenantId=dd8cbebb-2139-4df8-b411-4e3e87abeb5c&groupId=4620eac6-0955-4c69-93e8-b493b21f0df6&parentMessageId=1731692568019&teamName=ML%2F%20Cultural%20Heritage%20Support&channelName=1%20-%20LUX_ML&createdTime=1731692568019)

**Comment by zachjschwartz** [comment]
@kamerynB not sure if this still needs to be in forming—looks well-formed to me. I'll add to forming for tomorrow
